### PR TITLE
Make snippets/buildCache/cacheable-bundle-task compatible with the co…

### PR DESCRIPTION
**Pull request by DmitryPogrebnoy**
…nfiguration cache

The tasks in the snippet were accessing Task.getProject().file() at the execution
time. This usage was replaced with a script variable to capture the result of using project.file(file).

No need to adjust text in docs.

Signed-off-by: DmitryPogrebnoy <pogrebnoy.inc@gmail.com>

Fixes [#21801](https://github.com/gradle/gradle/issues/21801)

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle [hunk 1]
```diff
@@ -7,10 +7,15 @@ abstract class NpmTask extends DefaultTask {
 
     ListProperty<String> args = project.objects.listProperty(String)
 
+    @Internal
+    protected File bundleJSFile = project.file("$project.buildDir/bundle.js")
```
Comment (bamboo) [inline subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle:11]: 🔴 This changes semantics as the evaluation of `buildDir` would now happen at task instantiation time and could end up observing a stale value. 

An alternative solution would be to use an injected `ProjectLayout`.

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle [hunk 2]
```diff
@@ -7,10 +7,15 @@ abstract class NpmTask extends DefaultTask {
...
+    @Internal
+    protected File[] scriptsFiles = project.file("scripts").listFiles()
```
Comment (bamboo) [inline subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle:13]: 🔴 This changes semantics as the `listFiles()` call would now happen at task instantiation time and not at task execution time.

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle (hunk 1 above)
Comment (DmitryPogrebnoy) [reply subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle]: Ok, I see. Now I try to use injected `ProjectLayout` as you suggest and it's seems it's works. Recheck please

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle (hunk 2 above)
Comment (DmitryPogrebnoy) [reply subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle]: Sure, fixed I guess

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle [hunk 3]
```diff
@@ -7,10 +7,17 @@ abstract class NpmTask extends DefaultTask {
...
+    @Inject
+    abstract ProjectLayout getProjectLayout()
+
     @TaskAction
     void run() {
-        project.file("$project.buildDir/bundle.js").withOutputStream { stream ->
-            project.file("scripts").listFiles().sort().each {
+        String bundleFilePath = "${projectLayout.buildDirectory.asFile.get()}/bundle.js"
+        File bundleFile = projectLayout.projectDirectory.file(bundleFilePath).asFile
+        File[] scriptsFiles = projectLayout.projectDirectory.file("scripts").asFile.listFiles()
```
Comment (bamboo) [inline subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle:17]: 🔴 `dir("scripts")` would be more accurate, wouldn't it?

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle [hunk 4]
```diff
@@ -7,10 +7,17 @@ abstract class NpmTask extends DefaultTask {
...
 
+    @Inject
+    abstract ProjectLayout getProjectLayout()
+
     @TaskAction
     void run() {
-        project.file("$project.buildDir/bundle.js").withOutputStream { stream ->
-            project.file("scripts").listFiles().sort().each {
+        String bundleFilePath = "${projectLayout.buildDirectory.asFile.get()}/bundle.js"
+        File bundleFile = projectLayout.projectDirectory.file(bundleFilePath).asFile
```
[the remark this answers is not in the document]
Comment (DmitryPogrebnoy) [reply subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle]: Yeh, fixed

File: subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle (hunk 3 above)
Comment (DmitryPogrebnoy) [reply subprojects/docs/src/snippets/buildCache/cacheable-bundle-task/groovy/build.gradle]: Thanks for suggestion, fixed

**Comment by bamboo** [comment]
@bot-gradle test this

**Review by bamboo — APPROVED** [review]
:tada:

**Comment by bamboo** [comment]
@bot-gradle test and merge

**Pull request merged**